RFC 1: Max BTC Fee Revamped (#113)
keep-network/tbtc-v2: docs/rfc/rfc-1.adoc

[[bitcoin-sweeping-fee]]
=== Bitcoin Sweeping Fee
Governable Parameters

- `btc_fee_broadcast_timeout`: The amount of an operator has to provide a suggested BTC fee
  before the other operators give up and try the next operator.

Any time a bitcoin transaction needs to be posted and then mined on the bitcoin
blockchain, the miners need to be paid a fee for their work. This fee
fluctuates with market demand and is decently volatile.

The operators need to all agree on a fee before they can construct the sweeping

> transacion, but communicating that fee is tricky. When the sortition pool

Reviewer: ```suggestion
transaction, but communicating that fee is tricky. When the sortition pool
```

Author: addressed in 6a3ff94

selects the operators, it selects them in an ordered list. We can leverage this
on-chain order to be the fee-proposal-order. Say that the sortition pool chose
Operators: [#71, #109, #34..., #2]. When it comes time to sweep deposits,
Operator#71 would be expected to query
https://blockstream.info/api/fee-estimates for the 3-block fee and broadcast
this fee to the rest of the operators. If the operators don't receive the fee
from operator #71 before `btc_fee_broadcast_timeout`, has elapsed, the duty

> moves to #109, and then #34, and so on.

Reviewer 2: I think we should specify what is happening with `#71` and `#109`. Can they still propose a fee?

I would say no because then we could have a problem with 3 operators proposing different fees and which one to choose. 🤔

Author: I was under the assumption that it looks something like "okay it's block 25. operator #71, that means you have to give us a fee. you have until block 45 to do so, and then it'll be operator 109's turn"

The operators then validate the fee they received against
https://blockstream.info/api/fee-estimates to make sure that nothing fishy is
going on - that the fee isn't too high or too low (TBD what that means). If it

> is, they can 51-of-100 agree to skip to the next operator.

Reviewer 2: What does it mean they 51-of-100 agree? I thought that skipping the operator after a certain time passes is just a part of the algorithm they are following and they do not need to additionally agree (communicate) to skip the operator. 🤔

Author: Okay - talked a little bit about this offline, producing extra signatures for off-chain stuff runs into nonce-reuse problems: https://billatnapier.medium.com/ecdsa-weakness-where-nonces-are-reused-2be63856a01a . The pocket veto strategy (simply wait to skip a bad fee) works with less complication and less message exchange

Author: Okay - revised in 6a65271 

can you check that out and see if you like the phrasing there?

*Note*: We purposefully leave out on-chain incentives/punishments here. The
attack vector is small and the overhead is high. We might need to revisit this

> in the future, but if someone maintianed a client fork in order to take
> advanage of manipulating btc fee consensus I would be very surprised.

Reviewer: ```suggestion
in the future, but if someone maintained a client fork in order to take
advantage of manipulating btc fee consensus I would be very surprised.
```

Author: addressed in 6a3ff94

> advanage of manipulating btc fee consensus I would be very surprised.

Reviewer 2: All clients validate the fee independently. We use the mechanism of one client proposing the fee so that we do not have to "medianize" multiple fee propositions. 🤷🏻

== Governable Parameters
Alphabetized list of Governable Parameters with additional notes.

- `base_sweeping_fee_max`: The highest amount of BTC that operators can
  initially propose as a fee for miners in a sweeping transaction.

> - `btc_fee_broadcast_timeout`: The amount of an operator has to provide a suggested BTC fee

Reviewer: ```suggestion
- `btc_fee_broadcast_timeout`: The amount of time an operator has to provide a suggested BTC fee
```

Author: addressed in 95285dd
